<?xml version="1.0"?>
<!--
Copyright (c) 2020 Pilz GmbH & Co. KG

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
-->
<launch>

  <!-- Configure modbus settings depending on the configured safety hardware -->
  <arg name="safety_hw" default="pss4000"/>
  <arg name="read_api_spec_file" default="$(find prbt_hardware_support)/config/modbus_read_api_spec_$(arg safety_hw).yaml" />
  <arg name="write_api_spec_file" default="$(find prbt_hardware_support)/config/modbus_write_api_spec_$(arg safety_hw).yaml" />
  <arg name="modbus_server_ip" default="$(eval '192.168.0.10' if safety_hw=='pss4000' else '169.254.60.1')" />

  <!-- Options to enable and disable different hardware aspects -->
  <arg name="has_braketest_support" default="true"/>
  <arg name="has_operation_mode_support" default="true"/>

  <!-- Read modbus register specifications -->
  <rosparam ns="/prbt/read_api_spec" command="load" file="$(arg read_api_spec_file)" />
  <rosparam ns="/prbt/write_api_spec" command="load" file="$(arg write_api_spec_file)" if="$(arg has_braketest_support)" />

  <!-- Open modbus connection and required modules -->
  <!-- Modbus connection -->
  <include file="$(find prbt_hardware_support)/launch/modbus_client.launch">
    <arg name="modbus_server_ip" value="$(arg modbus_server_ip)" />
  </include>
  <!-- Run permitted -->
  <include file="$(find prbt_hardware_support)/launch/modbus_adapter_run_permitted_node.launch" />
  <include file="$(find prbt_hardware_support)/launch/stop1_executor_node.launch" />

  <!-- Brake test -->
  <include if="$(arg has_braketest_support)" file="$(find prbt_hardware_support)/launch/brake_test.launch">
    <arg name="write_api_spec_file" value="$(arg write_api_spec_file)" />
  </include>

  <!-- Operation Mode -->
  <include if="$(arg has_operation_mode_support)"
           file="$(find prbt_hardware_support)/launch/operation_mode.launch" />
  <node unless="$(arg has_operation_mode_support)"
        pkg="prbt_hardware_support" name="fake_speed_override_node" type="fake_speed_override_node"/>
</launch>
